tls/s2n_client_key_exchange.c

```diff
@@ -110,6 +111,23 @@ int s2n_ecdhe_client_key_recv(struct s2n_connection *conn, struct s2n_blob *shar
     return 0;
 }
 
+int s2n_kem_client_recv_key(struct s2n_connection *conn, struct s2n_blob *shared_key)
+{
+    struct s2n_stuffer *in = &conn->handshake.io;
+    kem_ciphertext_key_size ciphertext_length;
+
+    GUARD(s2n_stuffer_read_uint16(in, &ciphertext_length));
+    S2N_ERROR_IF(ciphertext_length > s2n_stuffer_data_available(in), S2N_ERR_BAD_MESSAGE);
+    const struct s2n_blob ciphertext = {.size = ciphertext_length, .data = s2n_stuffer_raw_read(in, ciphertext_length)};
+    notnull_check(ciphertext.data);
+
+    GUARD(s2n_kem_decapsulate(&conn->secure.kem_params, shared_key, &ciphertext));
+
+    GUARD(s2n_free(&conn->secure.kem_params.private_key));
```

Comment:
Why free the private_key here? Not meant to be a leading question. I'm just curious what the rationale is and I'm guessing it is related to how the kem-style keys are used in the flow of the handshake. Is there a single place where all of the kem_params can be cleaned up at once?

Reply:
After this point the server doesn't need it anymore and I thought it was good to get ride of any secret as soon as it is no longer needed. It also follows the pattern of `s2n_ecdhe_client_key_recv` calling `s2n_ecc_params_free` once it's done with the parameters (same with DHE). Because this is a KEM this is a key and not a half of the params so this should actually be wiped then freed. However it does make memory cleanup a little confusing because it can be freed here or in `s2n_kem_wipe_keys`.